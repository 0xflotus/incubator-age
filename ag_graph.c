#include "postgres.h"

#include "access/genam.h"
#include "access/heapam.h"
#include "access/htup.h"
#include "access/htup_details.h"
#include "access/skey.h"
#include "access/stratnum.h"
#include "catalog/indexing.h"
#include "storage/lockdefs.h"
#include "utils/fmgroids.h"
#include "utils/lsyscache.h"
#include "utils/rel.h"
#include "utils/relcache.h"

#include "ag_catalog.h"
#include "ag_graph.h"

#define Anum_ag_graph_name 1
#define Anum_ag_graph_namespace 2

#define Natts_ag_graph 2

static Oid ag_graph_relation_id(void)
{
    Oid id;

    id = get_relname_relid("ag_graph", ag_catalog_namespace_id());
    if (!OidIsValid(id)) {
        ereport(ERROR, (errcode(ERRCODE_UNDEFINED_TABLE),
                        errmsg("table \"ag_graph\" does not exist")));
    }

    return id;
}

static Oid ag_graph_name_index_id(void)
{
    Oid id;

    id = get_relname_relid("ag_graph_name_index", ag_catalog_namespace_id());
    if (!OidIsValid(id)) {
        ereport(ERROR,
                (errcode(ERRCODE_UNDEFINED_TABLE),
                 errmsg("index \"ag_graph_name_index\" does not exist")));
    }

    return id;
}

// INSERT INTO ag_catalog.ag_graph VALUES (graph_name, nsp_id)
void insert_graph(const Name graph_name, const Oid nsp_id)
{
    Datum values[Natts_ag_graph];
    bool nulls[Natts_ag_graph];
    Relation ag_graph;
    HeapTuple tuple;

    values[Anum_ag_graph_name - 1] = NameGetDatum(graph_name);
    nulls[Anum_ag_graph_name - 1] = false;

    values[Anum_ag_graph_namespace - 1] = ObjectIdGetDatum(nsp_id);
    nulls[Anum_ag_graph_namespace - 1] = false;

    ag_graph = heap_open(ag_graph_relation_id(), RowExclusiveLock);

    tuple = heap_form_tuple(RelationGetDescr(ag_graph), values, nulls);

    // CatalogTupleInsert() is originally for PostgreSQL's catalog. However,
    // it is used at here for convenience. We don't care about the returned OID
    // because ag_graph doesn't have OID column.
    CatalogTupleInsert(ag_graph, tuple);

    heap_close(ag_graph, RowExclusiveLock);
}

// DELETE FROM ag_catalog.ag_graph WHERE name = graph_name
void delete_graph(const Name graph_name)
{
    ScanKeyData scan_keys[1];
    Relation ag_graph;
    SysScanDesc scan_desc;
    HeapTuple tuple;

    ScanKeyInit(&scan_keys[0], Anum_ag_graph_name, BTEqualStrategyNumber,
                F_NAMEEQ, NameGetDatum(graph_name));

    ag_graph = heap_open(ag_graph_relation_id(), RowExclusiveLock);

    scan_desc = systable_beginscan(ag_graph, ag_graph_name_index_id(), true,
                                   NULL, 1, scan_keys);

    tuple = systable_getnext(scan_desc);
    if (!HeapTupleIsValid(tuple)) {
        ereport(ERROR,
                (errcode(ERRCODE_UNDEFINED_SCHEMA),
                 errmsg("graph \"%s\" does not exist", NameStr(*graph_name))));
    }

    CatalogTupleDelete(ag_graph, &tuple->t_self);

    systable_endscan(scan_desc);

    heap_close(ag_graph, RowExclusiveLock);
}

Oid get_graph_namespace(const Name graph_name)
{
    ScanKeyData scan_keys[1];
    Relation ag_graph;
    SysScanDesc scan_desc;
    HeapTuple tuple;
    bool is_null;
    Datum value;
    Oid nsp_id;

    ScanKeyInit(&scan_keys[0], Anum_ag_graph_name, BTEqualStrategyNumber,
                F_NAMEEQ, NameGetDatum(graph_name));

    ag_graph = heap_open(ag_graph_relation_id(), AccessShareLock);

    scan_desc = systable_beginscan(ag_graph, ag_graph_name_index_id(), true,
                                   NULL, 1, scan_keys);

    tuple = systable_getnext(scan_desc);
    if (!HeapTupleIsValid(tuple)) {
        ereport(ERROR,
                (errcode(ERRCODE_UNDEFINED_SCHEMA),
                 errmsg("graph \"%s\" does not exist", NameStr(*graph_name))));
    }

    value = fastgetattr(tuple, Anum_ag_graph_namespace,
                        RelationGetDescr(ag_graph), &is_null);
    if (is_null) {
        ereport(ERROR, (errmsg("namespace of graph \"%s\" is NULL",
                               NameStr(*graph_name))));
    }

    nsp_id = DatumGetObjectId(value);

    systable_endscan(scan_desc);

    heap_close(ag_graph, AccessShareLock);

    return nsp_id;
}
