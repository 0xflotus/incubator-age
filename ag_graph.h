#ifndef AG_AG_GRAPH_H
#define AG_AG_GRAPH_H

void insert_graph(const Name graph_name, const Oid nsp_id);
void delete_graph(const Name graph_name);

Oid get_graph_namespace(const Name graph_name);

#endif
