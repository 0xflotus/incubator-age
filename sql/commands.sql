LOAD 'agensgraph';
SET search_path TO ag_catalog;

SELECT create_graph('g');
SELECT count(*) FROM ag_graph WHERE name = 'g';
SELECT count(*) FROM pg_namespace WHERE nspname = 'g';

-- Create a temporary table to test drop_graph().
CREATE TABLE g.tmp (i int);

SELECT drop_graph('g');
SELECT drop_graph('g', true);
SELECT count(*) FROM pg_namespace WHERE nspname = 'g';
SELECT count(*) FROM ag_graph WHERE name = 'g';

SELECT create_graph(NULL);
SELECT drop_graph(NULL);
