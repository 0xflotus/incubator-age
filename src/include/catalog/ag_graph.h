#ifndef AG_AG_GRAPH_H
#define AG_AG_GRAPH_H

#include "postgres.h"

Oid insert_graph(const Name graph_name, const Oid nsp_id);
void delete_graph(const Name graph_name);
void update_graph_name(const Name graph_name, const Name new_name);

Oid get_graph_oid(const char *graph_name);
Oid get_graph_namespace(const char *graph_name);

#endif
