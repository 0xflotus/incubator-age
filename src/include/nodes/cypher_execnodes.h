#ifndef AG_CYPHER_SCAN_STATE_H
#define AG_CYPHER_SCAN_STATE_H

#include "executor/tuptable.h"
#include "nodes/extensible.h"
#include "nodes/relation.h"

typedef struct cypher_create_scan_state
{
    CustomScanState css;
    List *pattern;
} cypher_create_scan_state;

extern const CustomExecMethods cypher_create_custom_exec_methods;

#endif
