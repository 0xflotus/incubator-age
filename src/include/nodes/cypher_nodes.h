/*
 * Copyright 2020 Bitnine Co., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef AG_CYPHER_NODE_H
#define AG_CYPHER_NODE_H

#include "postgres.h"

#include "nodes/extensible.h"
#include "nodes/parsenodes.h"
#include "nodes/pg_list.h"

#include "nodes/ag_nodes.h"

/*
 * clauses
 */

typedef struct cypher_return
{
    ExtensibleNode extensible;
    bool distinct;
    List *items; // a list of ResTarget's
    List *order_by;
    Node *skip;
    Node *limit;
} cypher_return;

typedef struct cypher_with
{
    ExtensibleNode extensible;
    bool distinct;
    List *items; // a list of ResTarget's
    List *order_by;
    Node *skip;
    Node *limit;
    Node *where;
} cypher_with;

typedef struct cypher_match
{
    ExtensibleNode extensible;
    List *pattern; // a list of cypher_paths
    Node *where; // optional WHERE subclause (expression)
} cypher_match;

typedef struct cypher_create
{
    ExtensibleNode extensible;
    List *pattern; // a list of cypher_paths
} cypher_create;

typedef struct cypher_set
{
    ExtensibleNode extensible;
    List *items; // a list of cypher_set_items
    bool is_remove; // true if this is REMOVE clause
} cypher_set;

typedef struct cypher_set_item
{
    ExtensibleNode extensible;
    Node *prop; // LHS
    Node *expr; // RHS
    bool is_add; // true if this is +=
} cypher_set_item;

typedef struct cypher_delete
{
    ExtensibleNode extensible;
    bool detach; // true if DETACH is specified
    List *exprs; // targets of this deletion
} cypher_delete;

/*
 * pattern
 */

typedef struct cypher_path
{
    ExtensibleNode extensible;
    List *path; // [ node ( , relationship , node , ... ) ]
    int location;
} cypher_path;

// ( name :label props )
typedef struct cypher_node
{
    ExtensibleNode extensible;
    char *name;
    char *label;
    Node *props; // map or parameter
    int location;
} cypher_node;

typedef enum
{
    CYPHER_REL_DIR_NONE,
    CYPHER_REL_DIR_LEFT,
    CYPHER_REL_DIR_RIGHT
} cypher_rel_dir;

// -[ name :label props ]-
typedef struct cypher_relationship
{
    ExtensibleNode extensible;
    char *name;
    char *label;
    Node *props; // map or parameter
    cypher_rel_dir dir;
    int location;
} cypher_relationship;

/*
 * expression
 */

typedef struct cypher_bool_const
{
    ExtensibleNode extensible;
    bool boolean;
    int location;
} cypher_bool_const;

typedef struct cypher_param
{
    ExtensibleNode extensible;
    char *name;
    int location;
} cypher_param;

typedef struct cypher_map
{
    ExtensibleNode extensible;
    List *keyvals;
    int location;
} cypher_map;

typedef struct cypher_list
{
    ExtensibleNode extensible;
    List *elems;
    int location;
} cypher_list;

enum cypher_string_match_op
{
    CSMO_STARTS_WITH,
    CSMO_ENDS_WITH,
    CSMO_CONTAINS
};

typedef struct cypher_string_match
{
    ExtensibleNode extensible;
    enum cypher_string_match_op operation;
    Node *lhs;
    Node *rhs;
    int location;
} cypher_string_match;

typedef struct cypher_target_node
{
    ResultRelInfo *resultRelInfo;
    TupleTableSlot *elemTupleSlot;
    Oid relid;
    List *targetList;
    List *expr_states;
} cypher_target_node;

typedef struct cypher_typecast
{
    ExtensibleNode extensible;
    Node *expr;
    char *typecast;
    int location;
} cypher_typecast;

/* clauses */
void out_cypher_return(StringInfo str, const ExtensibleNode *node);
void out_cypher_with(StringInfo str, const ExtensibleNode *node);
void out_cypher_match(StringInfo str, const ExtensibleNode *node);
void out_cypher_create(StringInfo str, const ExtensibleNode *node);
void out_cypher_set(StringInfo str, const ExtensibleNode *node);
void out_cypher_set_item(StringInfo str, const ExtensibleNode *node);
void out_cypher_delete(StringInfo str, const ExtensibleNode *node);

/* pattern */
void out_cypher_path(StringInfo str, const ExtensibleNode *node);
void out_cypher_node(StringInfo str, const ExtensibleNode *node);
void out_cypher_relationship(StringInfo str, const ExtensibleNode *node);

/* expression */
void out_cypher_bool_const(StringInfo str, const ExtensibleNode *node);
void out_cypher_param(StringInfo str, const ExtensibleNode *node);
void out_cypher_map(StringInfo str, const ExtensibleNode *node);
void out_cypher_list(StringInfo str, const ExtensibleNode *node);

/* string match */
void out_cypher_string_match(StringInfo str, const ExtensibleNode *node);

#endif
