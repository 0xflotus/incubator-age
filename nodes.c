#include "postgres.h"

#include "nodes/extensible.h"

#include "cypher_nodes.h"
#include "nodes.h"

static void copy_ag_node(ExtensibleNode *newnode,
                         const ExtensibleNode *oldnode);
static bool equal_ag_node(const ExtensibleNode *a, const ExtensibleNode *b);
static void read_ag_node(ExtensibleNode *node);

// This list must match ag_node_tag.
const char *node_names[] = {
    "cypher_return",
    "cypher_with",
    "cypher_set_clause",
    "cypher_set_prop"
};

#define DEFINE_NODE_METHODS(type) \
    { \
        #type, \
        sizeof(type), \
        copy_ag_node, \
        equal_ag_node, \
        out_##type, \
        read_ag_node \
    }

const ExtensibleNodeMethods node_methods[] = {
    DEFINE_NODE_METHODS(cypher_return),
    DEFINE_NODE_METHODS(cypher_with),
    DEFINE_NODE_METHODS(cypher_set_clause),
    DEFINE_NODE_METHODS(cypher_set_prop)
};

static void copy_ag_node(ExtensibleNode *newnode,
                         const ExtensibleNode *oldnode)
{
    ereport(ERROR, (errmsg("unexpected copyObject() over ag_node")));
}

static bool equal_ag_node(const ExtensibleNode *a, const ExtensibleNode *b)
{
    ereport(ERROR, (errmsg("unexpected equal() over ag_node's")));
}

static void read_ag_node(ExtensibleNode *node)
{
    ereport(ERROR, (errmsg("unexpected parseNodeString() for ag_node")));
}

void register_ag_nodes(void)
{
    int i;

    for (i = 0; i < lengthof(node_methods); i++)
        RegisterExtensibleNodeMethods(&node_methods[i]);
}

ExtensibleNode *new_ag_node(Size size, ag_node_tag tag)
{
    ExtensibleNode *n;

    n = (ExtensibleNode *)palloc0fast(size);
    n->type = T_ExtensibleNode;
    n->extnodename = node_names[tag];

    return n;
}
