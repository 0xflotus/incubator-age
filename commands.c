#include "postgres.h"

#include "access/xact.h"
#include "catalog/dependency.h"
#include "catalog/objectaddress.h"
#include "catalog/pg_namespace.h"
#include "commands/schemacmds.h"
#include "fmgr.h"
#include "miscadmin.h"
#include "nodes/nodes.h"
#include "nodes/parsenodes.h"
#include "utils/acl.h"
#include "utils/lsyscache.h"

#include "ag_graph.h"

static Oid create_schema_for_graph(const Name graph_name);
static void drop_schema_for_graph(const Name graph_name, const bool cascade);

PG_FUNCTION_INFO_V1(create_graph);

Datum create_graph(PG_FUNCTION_ARGS)
{
    Name graph_name;
    Oid nsp_id;

    if (PG_ARGISNULL(0)) {
        ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                        errmsg("graph name must not be NULL")));
    }
    graph_name = PG_GETARG_NAME(0);

    nsp_id = create_schema_for_graph(graph_name);
    insert_graph(graph_name, nsp_id);

    // Make sure later steps can see the object created here.
    CommandCounterIncrement();

    ereport(NOTICE,
            (errmsg("graph \"%s\" has been created", NameStr(*graph_name))));

    PG_RETURN_VOID();
}

static Oid create_schema_for_graph(const Name graph_name)
{
    CreateSchemaStmt *stmt;
    Oid nsp_id;

    // This is the same with running `CREATE SCHEMA graph_name`.
    // schemaname doesn't have to be graph_name but the same name is used so
    // that users can find the backed schema for a graph only by its name.
    stmt = makeNode(CreateSchemaStmt);
    stmt->schemaname = NameStr(*graph_name);
    stmt->authrole = NULL;
    stmt->schemaElts = NIL;
    stmt->if_not_exists = false;
    nsp_id = CreateSchemaCommand(stmt, "(generated CREATE SCHEMA command)", -1,
                                 -1);

    return nsp_id;
}

PG_FUNCTION_INFO_V1(drop_graph);

Datum drop_graph(PG_FUNCTION_ARGS)
{
    Name graph_name;
    bool cascade;

    if (PG_ARGISNULL(0)) {
        ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                        errmsg("graph name must not be NULL")));
    }
    graph_name = PG_GETARG_NAME(0);
    cascade = PG_GETARG_BOOL(1);

    drop_schema_for_graph(graph_name, cascade);
    delete_graph(graph_name);

    // Make effects of this command visible.
    CommandCounterIncrement();

    ereport(NOTICE,
            (errmsg("graph \"%s\" has been dropped", NameStr(*graph_name))));

    PG_RETURN_VOID();
}

static void drop_schema_for_graph(const Name graph_name, const bool cascade)
{
    Oid nsp_id;
    ObjectAddress object;
    DropBehavior behavior;

    nsp_id = get_graph_namespace(graph_name);

    if (!pg_namespace_ownercheck(nsp_id, GetUserId())) {
        aclcheck_error(ACLCHECK_NOT_OWNER, OBJECT_SCHEMA,
                       get_namespace_name(nsp_id));
    }

    object.classId = NamespaceRelationId;
    object.objectId = nsp_id;
    object.objectSubId = 0;

    behavior = cascade ? DROP_CASCADE : DROP_RESTRICT;

    performDeletion(&object, behavior, 0);
}
