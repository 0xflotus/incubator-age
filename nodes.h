#ifndef AG_NODE_H
#define AG_NODE_H

#include "postgres.h"

#include "nodes/extensible.h"
#include "nodes/nodes.h"

typedef enum ag_node_tag
{
    // projection
    cypher_return_t,
    cypher_with_t,
    // reading clause
    cypher_match_t,
    // updating clause
    cypher_create_t,
    cypher_set_t,
    cypher_set_item_t,
    cypher_delete_t,
    // pattern
    cypher_path_t,
    cypher_node_t,
    cypher_relationship_t
} ag_node_tag;

void register_ag_nodes(void);

ExtensibleNode *new_ag_node(Size size, ag_node_tag tag);
#define make_ag_node(type) \
    ((type *)new_ag_node(sizeof(type), CppConcat(type, _t)))

#define is_ag_node(node, type) _is_ag_node(node, #type)
static inline bool _is_ag_node(Node *node, const char *extnodename)
{
    ExtensibleNode *extnode;

    if (!IsA(node, ExtensibleNode))
        return false;

    extnode = (ExtensibleNode *)node;
    if (strcmp(extnode->extnodename, extnodename) == 0)
        return true;

    return false;
}

#endif
