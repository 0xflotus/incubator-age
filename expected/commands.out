LOAD 'agensgraph';
SET search_path TO ag_catalog;
SELECT create_graph('g');
NOTICE:  graph "g" has been created
 create_graph 
--------------
 
(1 row)

SELECT count(*) FROM ag_graph WHERE name = 'g';
 count 
-------
     1
(1 row)

SELECT count(*) FROM pg_namespace WHERE nspname = 'g';
 count 
-------
     1
(1 row)

-- Create a temporary table to test drop_graph().
CREATE TABLE g.tmp (i int);
SELECT drop_graph('g');
ERROR:  cannot drop schema g because other objects depend on it
DETAIL:  table g.tmp depends on schema g
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
SELECT drop_graph('g', true);
NOTICE:  drop cascades to table g.tmp
NOTICE:  graph "g" has been dropped
 drop_graph 
------------
 
(1 row)

SELECT count(*) FROM pg_namespace WHERE nspname = 'g';
 count 
-------
     0
(1 row)

SELECT count(*) FROM ag_graph WHERE name = 'g';
 count 
-------
     0
(1 row)

SELECT create_graph(NULL);
ERROR:  graph name must not be NULL
SELECT drop_graph(NULL);
ERROR:  graph name must not be NULL
